src/stream/src/executor/row_merge.rs

```diff
@@ -0,0 +1,153 @@
+// Copyright 2024 RisingWave Labs
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use risingwave_common::bail;
+use risingwave_common::types::ToOwnedDatum;
+use risingwave_common::util::chunk_coalesce::DataChunkBuilder;
+use risingwave_common::util::column_index_mapping::ColIndexMapping;
+
+use super::barrier_align::*;
+use crate::executor::prelude::*;
+
+pub struct RowMergeExecutor {
+    ctx: ActorContextRef,
+    pub lhs_input: Executor,
+    pub rhs_input: Executor,
+    /// Maps input from the lhs to the output.
+    pub lhs_mapping: ColIndexMapping,
+    /// Maps input from the rhs to the output.
+    pub rhs_mapping: ColIndexMapping,
+    /// Output schema
+    pub schema: Schema,
+}
+
+impl RowMergeExecutor {
+    pub fn new(
+        ctx: ActorContextRef,
+        lhs_input: Executor,
+        rhs_input: Executor,
+        lhs_mapping: ColIndexMapping,
+        rhs_mapping: ColIndexMapping,
+        schema: Schema,
+    ) -> Self {
+        Self {
+            ctx,
+            lhs_input,
+            rhs_input,
+            lhs_mapping,
+            rhs_mapping,
+            schema,
+        }
+    }
+
+    #[try_stream(ok = Message, error = StreamExecutorError)]
+    pub async fn execute_inner(self) {
+        let lhs_mapping = self.lhs_mapping;
+        let rhs_mapping = self.rhs_mapping;
+        let data_types = self
+            .schema
+            .fields()
+            .iter()
+            .map(|f| f.data_type())
+            .collect::<Vec<_>>();
+
+        {
+            let mut lhs_buffer: Option<StreamChunk> = None;
+            let mut rhs_buffer: Option<StreamChunk> = None;
+            let aligned_stream = barrier_align(
+                self.lhs_input.execute(),
+                self.rhs_input.execute(),
+                self.ctx.id,
+                self.ctx.fragment_id,
+                self.ctx.streaming_metrics.clone(),
+                "Join",
+            );
+            pin_mut!(aligned_stream);
+            #[for_await]
+            for message in aligned_stream {
+                match message? {
+                    AlignedMessage::Left(chunk) => {
+                        lhs_buffer = Some(chunk);
+                    }
+                    AlignedMessage::Right(chunk) => {
+                        rhs_buffer = Some(chunk);
+                    }
```

Comment:
This executor looks like a very special executor which accepts one chunk from each input per barrier. In my mind, it should be a general merge executor that can process any number of chunks from input and need to maintain a hash map to match rows for both side.

Reply:
I think it's similar to the difference between hash agg and simple agg. Once we decide to implement 'keyed merge', we have to consider whether the number of rows coming from both sides will exceed memory usage, and then consider whether to introduce a separate state table or spill mechanism. Therefore, introducing a simple implementation to determine the cardinality of both sides of the relation as 1 can quickly support percentile aggregation.

Reply:
Prefer a simpler executor compared to a more generalized one. We can generalize it if the need arises.

Reply:
> Got it. The row merge is the same as the simple agg and the generalized one should be called key merge. BTW, can we add an assertion to ensure lhs_buffer and rhs_buffer is None before assigning?

Hmm I think the normal agg executor may output multiple chunks in one epoch. I will instead parse the chunks to ensure consistent operations.

Reply:
Done in https://github.com/risingwavelabs/risingwave/pull/17930/commits/677edab74870da69ee9ff4374e6c28f5da688de9